File: server/src/main/java/com/uiptv/db/PublishedM3uChannelSelectionDb.java
package com.uiptv.db;

import com.uiptv.model.PublishedM3uChannelSelection;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import static com.uiptv.db.DatabaseUtils.DbTable.PUBLISHED_M3U_CHANNEL_SELECTION_TABLE;
import static com.uiptv.db.DatabaseUtils.insertTableSql;
import static com.uiptv.db.SQLConnection.connect;
import static com.uiptv.util.StringUtils.isBlank;

@SuppressWarnings("java:S6548")
public class PublishedM3uChannelSelectionDb extends BaseDb {
    private static PublishedM3uChannelSelectionDb instance;

    private PublishedM3uChannelSelectionDb() {
        super(PUBLISHED_M3U_CHANNEL_SELECTION_TABLE);
    }

    public static synchronized PublishedM3uChannelSelectionDb get() {
        if (instance == null) {
            instance = new PublishedM3uChannelSelectionDb();
        }
        return instance;
    }

    @Override
    PublishedM3uChannelSelection populate(ResultSet resultSet) {
        PublishedM3uChannelSelection selection = new PublishedM3uChannelSelection();
        selection.setDbId(nullSafeString(resultSet, "id"));
        selection.setAccountId(nullSafeString(resultSet, "accountId"));
        selection.setCategoryName(nullSafeString(resultSet, "categoryName"));
        selection.setChannelId(nullSafeString(resultSet, "channelId"));
        selection.setSelected(safeBoolean(resultSet, "selected"));
        return selection;
    }

    public List<PublishedM3uChannelSelection> getAllSelections() {
        return getAll(" ORDER BY id", new String[]{});
    }

    public void replaceSelections(Connection conn, List<PublishedM3uChannelSelection> selections) throws SQLException {
        try (PreparedStatement delete = conn.prepareStatement("DELETE FROM " + PUBLISHED_M3U_CHANNEL_SELECTION_TABLE.getTableName())) {
            delete.executeUpdate();
        }
        if (selections == null || selections.isEmpty()) {
            return;
        }
        try (PreparedStatement insert = conn.prepareStatement(insertTableSql(PUBLISHED_M3U_CHANNEL_SELECTION_TABLE))) {
            for (PublishedM3uChannelSelection selection : selections) {
                if (selection == null
                        || isBlank(selection.getAccountId())
                        || isBlank(selection.getCategoryName())
                        || isBlank(selection.getChannelId())) {
                    continue;
                }
                insert.setString(1, selection.getAccountId());
                insert.setString(2, selection.getCategoryName());
                insert.setString(3, selection.getChannelId());
                insert.setString(4, selection.isSelected() ? "1" : "0");
                insert.addBatch();
            }
            insert.executeBatch();
        }
    }

    public void deleteByAccountId(String accountId) {
        if (isBlank(accountId)) {
            return;
        }
        String sql = "DELETE FROM " + PUBLISHED_M3U_CHANNEL_SELECTION_TABLE.getTableName() + " WHERE accountId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.executeUpdate();
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to delete published M3U channel selection", e);
        }
    }
}

File: server/src/main/kotlin/com/uiptv/db/PublishedM3uChannelSelectionDb.kt
package com.uiptv.db

import com.uiptv.model.PublishedM3uChannelSelection
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SortOrder
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.Table
import org.jetbrains.exposed.sql.deleteAll
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll

class PublishedM3uChannelSelectionDb private constructor() : ExposedCrudRepository<String, PublishedM3uChannelSelection>() {
    companion object {
        private val instance = PublishedM3uChannelSelectionDb()

        @JvmStatic
        fun get(): PublishedM3uChannelSelectionDb = instance
    }

    override fun findAll(): List<PublishedM3uChannelSelection> = query {
        PublishedM3uChannelSelectionTable.selectAll()
            .orderBy(PublishedM3uChannelSelectionTable.id to SortOrder.ASC)
            .map(ResultRow::toPublishedM3uChannelSelection)
    }

    override fun findById(id: String): PublishedM3uChannelSelection? = query {
        id.toIntOrNull()
            ?.let { dbId -> PublishedM3uChannelSelectionTable.selectAll().where { PublishedM3uChannelSelectionTable.id eq dbId }.firstOrNull() }
            ?.toPublishedM3uChannelSelection()
    }

    override fun save(entity: PublishedM3uChannelSelection): PublishedM3uChannelSelection {
        query {
            val insertedId = PublishedM3uChannelSelectionTable.insert { row ->
                row[accountId] = entity.accountId.orEmpty()
                row[categoryName] = entity.categoryName.orEmpty()
                row[channelId] = entity.channelId.orEmpty()
                row[selected] = entity.selected.asDbBoolean()
            }[PublishedM3uChannelSelectionTable.id]
            entity.dbId = insertedId.toString()
        }
        return entity
    }

    override fun deleteById(id: String) {
        val dbId = id.toIntOrNull() ?: return
        query {
            PublishedM3uChannelSelectionTable.deleteWhere { PublishedM3uChannelSelectionTable.id eq dbId }
        }
    }

    fun getAllSelections(): List<PublishedM3uChannelSelection> = findAll()

    fun replaceSelections(selections: List<PublishedM3uChannelSelection>?) {
        query {
            replaceSelectionsInTransaction(selections)
        }
    }

    internal fun replaceSelectionsInTransaction(selections: List<PublishedM3uChannelSelection>?) {
        PublishedM3uChannelSelectionTable.deleteAll()
        selections.orEmpty()
            .filter { !it.accountId.isNullOrBlank() && !it.categoryName.isNullOrBlank() && !it.channelId.isNullOrBlank() }
            .forEach { selection ->
                PublishedM3uChannelSelectionTable.insert { row ->
                    row[accountId] = selection.accountId.orEmpty()
                    row[categoryName] = selection.categoryName.orEmpty()
                    row[channelId] = selection.channelId.orEmpty()
                    row[selected] = selection.selected.asDbBoolean()
                }
            }
    }

    fun deleteByAccountId(accountId: String?) {
        if (accountId.isNullOrBlank()) {
            return
        }
        query {
            PublishedM3uChannelSelectionTable.deleteWhere { PublishedM3uChannelSelectionTable.accountId eq accountId }
        }
    }
}

internal object PublishedM3uChannelSelectionTable : Table(DatabaseUtils.DbTable.PUBLISHED_M3U_CHANNEL_SELECTION_TABLE.tableName) {
    val id = integer("id").autoIncrement()
    val accountId = text("accountId")
    val categoryName = text("categoryName")
    val channelId = text("channelId")
    val selected = text("selected").nullable()

    override val primaryKey = PrimaryKey(id)
}

private fun ResultRow.toPublishedM3uChannelSelection(): PublishedM3uChannelSelection =
    PublishedM3uChannelSelection(
        dbId = this[PublishedM3uChannelSelectionTable.id].toString(),
        accountId = this[PublishedM3uChannelSelectionTable.accountId],
        categoryName = this[PublishedM3uChannelSelectionTable.categoryName],
        channelId = this[PublishedM3uChannelSelectionTable.channelId],
        selected = this[PublishedM3uChannelSelectionTable.selected] == "1"
    )
